File: 144. Binary Tree Preorder Traversal/Python/Solution.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None


class Solution(object):
    def preorderTraversal(self, root):
        """
        :type root: TreeNode
        :rtype: List[int]
        """
        stack = []
        result = []
        while root or stack:
            if root:
                result.append(root.val)
                stack.append(root)
                root = root.left
            else:
                node = stack.pop()
                root = node.right
        return result

File: 144. Binary Tree Preorder Traversal/C++/Solution.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    vector<int> preorderTraversal(TreeNode *root) {
        stack<TreeNode *> sta;
        vector<int> result;
        while (root != nullptr || !sta.empty()) {
            if (root != nullptr) {
                result.push_back(root->val);
                sta.push(root);
                root = root->left;
            } else {
                root = sta.top()->right;
                sta.pop();
            }
        }
        return result;
    }
};